File: Codeforces/Educational Codeforces Round 37 (Rated for Div. 2)/C++/E.cpp
/*
Author: Dung Tuan Le
University of Rochester
*/

#include <bits/stdc++.h>
#define fio ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL)
#define FOR(i, l, r) for (int i=l; i<=r; i++)
#define REP(i, r, l) for (int i=r; i>=l; i--)
#define pb push_back
#define sz(x) int(x.size())

using namespace std;

#define FORALL(iter, x) for (auto iter = x.begin(); iter != x.end(); iter++)
#define maxn 200000 + 10
long n, m, cnt;
set<long> s2;
set<long> notEdge[maxn];
vector<long> temp;

void visit(long s) {
  if (s2.find(s) != s2.end()) s2.erase(s2.find(s));
  queue<long> q;
  q.push(s);
  while (!q.empty()) {
    long u = q.front();
    q.pop();
    temp.clear();
    cnt++;
    FORALL(iter, s2) {
      long v = (*iter);
      if (notEdge[u].find(v) == notEdge[u].end()) {
        temp.pb(v);
        q.push(v);
      }
    }
    FOR(i, 0, sz(temp) - 1) if (s2.find(temp[i]) != s2.end()) s2.erase(s2.find(temp[i]));
  }
}

int main() {
	fio;
  cin >> n >> m;
  FOR(i, 1, m) {
    long u, v;
    cin >> u >> v;
    notEdge[u].insert(v);
    notEdge[v].insert(u);
  }
  FOR(i, 1, n) s2.insert(i);
  vector<long> res;
  FOR(i, 1, n) if (s2.find(i) != s2.end()) {
    cnt = 0;
    visit(i);
    res.pb(cnt);
  }
  sort(res.begin(), res.end());
  cout << sz(res) << '\n';
  FOR(i, 0, sz(res) - 1) cout << res[i] << " ";
	return 0;
}

File: Codeforces/Educational Codeforces Round 37 (Rated for Div. 2)/Python/E.py
from sys import stdout, exit
from operator import itemgetter, attrgetter, methodcaller
import Queue

maxn = 200000 + 10
noEdge = [set() for i in range(maxn)] # create an array of a empty set
notVisited = set() # create a empty set

def visit(s):
    notVisited.discard(s)
    stack = []
    stack.append(s)
    cnt = 1
    while len(stack) > 0:
        u = stack.pop()
        temp = []
        for v in notVisited:
            if v not in noEdge[u]:
                temp.append(v)
                stack.append(v)
        cnt += len(temp)
        for v in temp: notVisited.discard(v)

    return cnt

n, m = map(int, raw_input().split(' '))
for i in range(m):
    u, v = map(int, raw_input().split(' '))
    noEdge[u].add(v)
    noEdge[v].add(u)

res = []
for i in range(n): notVisited.add(i + 1)
current = 1
while len(notVisited) > 0:
    while (current not in notVisited) and (current <= n):
        current += 1
    if current > n:
        break
    res.append(visit(current))

res.sort()
print len(res)
print ' '.join(map(str, res))
